turbine/src/quic_endpoint.rs

```diff
@@ -182,52 +183,77 @@ async fn run_client(
 }
 
 async fn handle_connecting_error(
+    endpoint: Endpoint,
     connecting: Connecting,
     sender: Sender<(Pubkey, SocketAddr, Bytes)>,
     cache: Arc<RwLock<ConnectionCache>>,
 ) {
-    if let Err(err) = handle_connecting(connecting, sender, cache).await {
+    if let Err(err) = handle_connecting(endpoint, connecting, sender, cache).await {
         error!("handle_connecting: {err:?}");
     }
 }
 
 async fn handle_connecting(
+    endpoint: Endpoint,
     connecting: Connecting,
     sender: Sender<(Pubkey, SocketAddr, Bytes)>,
     cache: Arc<RwLock<ConnectionCache>>,
 ) -> Result<(), Error> {
     let connection = connecting.await?;
     let remote_address = connection.remote_address();
     let remote_pubkey = get_remote_pubkey(&connection)?;
-    handle_connection_error(remote_address, remote_pubkey, connection, sender, cache).await;
+    handle_connection_error(
+        endpoint,
+        remote_address,
+        remote_pubkey,
+        connection,
+        sender,
+        cache,
+    )
+    .await;
     Ok(())
 }
 
 async fn handle_connection_error(
+    endpoint: Endpoint,
     remote_address: SocketAddr,
     remote_pubkey: Pubkey,
     connection: Connection,
     sender: Sender<(Pubkey, SocketAddr, Bytes)>,
     cache: Arc<RwLock<ConnectionCache>>,
 ) {
     cache_connection(remote_address, remote_pubkey, connection.clone(), &cache).await;
-    if let Err(err) = handle_connection(remote_address, remote_pubkey, &connection, sender).await {
+    if let Err(err) = handle_connection(
+        &endpoint,
+        remote_address,
+        remote_pubkey,
+        &connection,
+        &sender,
+    )
+    .await
+    {
         drop_connection(remote_address, remote_pubkey, &connection, &cache).await;
         error!("handle_connection: {remote_pubkey}, {remote_address}, {err:?}");
     }
 }
 
 async fn handle_connection(
+    endpoint: &Endpoint,
     remote_address: SocketAddr,
     remote_pubkey: Pubkey,
     connection: &Connection,
-    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    sender: &Sender<(Pubkey, SocketAddr, Bytes)>,
 ) -> Result<(), Error> {
     // Assert that send won't block.
     debug_assert_eq!(sender.capacity(), None);
     loop {
         match connection.read_datagram().await {
-            Ok(bytes) => sender.send((remote_pubkey, remote_address, bytes))?,
+            Ok(bytes) => {
+                if let Err(err) = sender.send((remote_pubkey, remote_address, bytes)) {
+                    close_quic_endpoint(endpoint);
```

Comment:
Did you notice an issue from this?

Reply:
`sender.send` will error only if the channel is disconnected, which means that the other end of the channel is shut down; at which point there is no point in running the endpoint any longer and we want the endpoint to shut down as well.